### DISCUSSION TITLE
关于绘制窗口大小变化的图

### DISCUSSION THREAD

**Discussion by Taoist-G** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#73] [Q&A]
请问打印成这个样子可以吗？
![window_size3](https://user-images.githubusercontent.com/98197403/211832062-ed0a0829-7998-4099-9c33-392ab81d907f.png)

**Comment by TANSixu** [comment]
Hi，方便增加一些对这个图的说明嘛

**Reply by Taoist-G** [reply]
横纵坐标分别是时间和窗口大小，然后上面的点就是测出来的值。

**Reply by TANSixu** [reply]
> 横纵坐标分别是时间和窗口大小，然后上面的点就是测出来的值。

请问这个测出的值是指？以及方便描述一下那个阶段是slow start，congestion avoidance对应的阶段嘛

**Reply by Taoist-G** [reply]
测出的值指的是对应的时间和窗口大小。点分布的比较稀疏（增长较快）的地方是慢启动和快速恢复，点分布比较密集（增长较慢）的地方是拥塞避免。

**Reply by TANSixu** [reply]
> 测出的值指的是对应的时间和窗口大小。点分布的比较稀疏（增长较快）的地方是慢启动和快速恢复，点分布比较密集（增长较慢）的地方是拥塞避免。

大概理解了。所以这个图里后面几次高峰都是快速恢复，然后ssthresh很早就到2了，所以每次高峰之后都对应着从很小开始的拥塞避免。请问我的理解正确嘛？

**Reply by Taoist-G** [reply]
高峰之后应该是慢启动

**Reply by Taoist-G** [reply]
慢启动之后很快拥塞避免

**Reply by TANSixu** [reply]
> 慢启动之后很快拥塞避免

可是快速恢复发生超时才会去慢启动。如果高峰之后的慢启动是超时引起的话，为什么他们在慢启动后进入接下来拥塞避免时会从很小的数值开始呢（看起来比慢启动的结束位置要低很多）。

**Reply by Taoist-G** [reply]
文档中的图中的拥塞避免开始时的窗口大小似乎也比上一次慢启动结束位置低很多。

**Reply by TANSixu** [reply]
> 文档中的图中的拥塞避免开始时的窗口大小似乎也比上一次慢启动结束位置低很多。

文档中的图有两次慢启动，如图
<img width="677" alt="image" src="https://user-images.githubusercontent.com/66773369/211854693-945143bc-dc08-4ae1-bb22-09539030770d.png">

每次慢启动后面的拥塞避免是接着慢启动走的，没有变低诶

**Reply by Taoist-G** [reply]
![IX4BU9O$P1@VM6B8%R4S857](https://user-images.githubusercontent.com/98197403/211855967-6fedb8c7-c344-4b80-8021-0e62d39db059.png)
这里好像是慢启动。。。。

**Reply by TANSixu** [reply]
> 这里好像是慢启动。。。。

那请问这里的慢启动前面那个小一点的峰是什么呢，如图
<img width="58" alt="image" src="https://user-images.githubusercontent.com/66773369/211856529-2421645d-428e-44b4-83cc-3a5c1b96798d.png">

**Reply by Taoist-G** [reply]
慢启动或者快速恢复的时候超时了？（猜想）

**Reply by Taoist-G** [reply]
emmm，我们是按照这个FSM来写的。
![HO} 4KQ7IBW9 _} ( T8EQI](https://user-images.githubusercontent.com/98197403/211856937-b5df5dab-07ad-4697-a5ab-8936abbf00e3.png)

**Reply by TANSixu** [reply]
是的，状态机没问题，但是可能需要解释一下细节（fast recovery细节稍微多一些）。其他都非常不错，除了上述中间那一段我感觉稍有些问题。或许这是在一个复杂环境生成的，您方便在transfer test（只会固定丢一个包）中生成一个这样的cwnd变化图嘛，可以帮助我们更好的分析您的实现

**Reply by Taoist-G** [reply]
这个是在transfer test中生成的图
![window_size2 (1)](https://user-images.githubusercontent.com/98197403/211962154-7e9025a8-1e77-4961-984a-d64fd1800f06.png)

**Reply by Taoist-G** [reply]
它现在自动变得正常了。。。。
![window_size2 (2)](https://user-images.githubusercontent.com/98197403/211966100-2cfac174-8af4-4c0e-a2cd-fbfa36b2b2af.png)

**Reply by Taoist-G** [reply]
basic_concurrency_test的图：
![window_size2 (3)](https://user-images.githubusercontent.com/98197403/211966429-78ac0aec-dc92-4ebb-ac91-67bfbc6563ae.png)
advanced_test的图：
![window_size12](https://user-images.githubusercontent.com/98197403/211966467-97760a1d-30d6-4475-9bc7-03a10544e20d.png)

**Reply by TANSixu** [reply]
> 这个是在transfer test中生成的图 ![window_size2 (1)](https://user-

看起来好像发生了多次丢包，请问是发生了超时嘛？

**Reply by Taoist-G** [reply]
看上去每一段都是前一段指数级增长，后一段每次+1，所以推测是在慢启动和快速恢复之间来回横跳，应该是不停收到dupACK，最后超时

**Comment by oliver-zhang987** [comment]
请问可以使用tensorboard来绘制吗？

**Reply by TANSixu** [reply]
Hi，可以的，绘图只会在答辩时检查。但是请注意在提交的代码中去掉相关代码，因为测试环境中没有相关包。

**Reply by oliver-zhang987** [reply]
好的，谢谢！
